File: old/src/http-server.ts
#!/usr/bin/env node

/**
 * HTTP Server wrapper for MCP-Eyes
 *
 * Run this directly from your terminal to inherit accessibility permissions.
 * Usage: npx ts-node src/http-server.ts
 * Or after build: node dist/http-server.js
 *
 * SECURITY:
 * - Requires API key authentication via MCP_EYES_API_KEY environment variable
 * - Binds to localhost only by default (set MCP_EYES_HOST=0.0.0.0 to expose)
 * - All requests must include Authorization: Bearer <api-key> header
 *
 * This server exposes MCP-Eyes functionality via HTTP endpoints,
 * bypassing the permission issues when running as an MCP subprocess.
 */

import http from 'http';
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { URL } from 'url';
// @ts-ignore
import screenshotDesktop from 'screenshot-desktop';
import { mouse, Button, keyboard, Key } from '@nut-tree-fork/nut-js';
import { run, checkAccessibilityPermissions, runAppleScript } from './jxa-runner.js';
import sharp from 'sharp';
import { AppleWindowManager } from './apple-window-manager.js';
import { OCRAnalyzer } from './ocr-analyzer.js';
import { LocalLLMAnalyzer } from './local-llm-analyzer.js';

declare const Application: any;

interface AppInfo {
  name: string;
  bundleId: string;
  pid: number;
  bounds: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
}

const PORT = parseInt(process.env.MCP_EYES_PORT || '3456');
const HOST = process.env.MCP_EYES_HOST || '127.0.0.1'; // localhost only by default
const TOKEN_FILE = path.join(process.env.HOME || '/tmp', '.mcp-eyes-token');

// Generate or load API key
function getOrCreateApiKey(): string {
  // If provided via environment, use that
  if (process.env.MCP_EYES_API_KEY) {
    return process.env.MCP_EYES_API_KEY;
  }

  // Try to load existing key from token file
  try {
    if (fs.existsSync(TOKEN_FILE)) {
      const content = fs.readFileSync(TOKEN_FILE, 'utf-8');
      const existing = JSON.parse(content);
      if (existing.apiKey && existing.port === PORT && existing.host === HOST) {
        // Reuse existing key if port/host match
        console.log('📌 Reusing existing API key from token file');
        return existing.apiKey;
      }
    }
  } catch (err) {
    // Token file doesn't exist or is invalid, generate new key
  }

  // Generate a new random key
  const apiKey = crypto.randomBytes(32).toString('hex');

  // Save to token file for MCP client to read
  try {
    fs.writeFileSync(TOKEN_FILE, JSON.stringify({
      apiKey,
      port: PORT,
      host: HOST,
      createdAt: new Date().toISOString()
    }), { mode: 0o600 }); // Only owner can read
  } catch (err) {
    console.error('Warning: Could not save token file:', err);
  }

  return apiKey;
}

class MCPEyesHTTPServer {
  private appleWindowManager: AppleWindowManager;
  private ocrAnalyzer: OCRAnalyzer;
  private localLLMAnalyzer: LocalLLMAnalyzer;
  private currentApp: AppInfo | null = null;
  private apiKey: string;

  constructor() {
    this.appleWindowManager = new AppleWindowManager();
    this.ocrAnalyzer = new OCRAnalyzer();
    this.localLLMAnalyzer = new LocalLLMAnalyzer();
    this.apiKey = getOrCreateApiKey();
  }

  async listApplications(): Promise<AppInfo[]> {
    // Use AppleScript which works when run from terminal
    const script = `
      tell application "System Events"
        set appList to ""
        repeat with proc in (every application process)
          set procName to name of proc
          set procBundle to bundle identifier of proc
          set procPID to unix id of proc
          set winBounds to "0,0,0,0"
          try
            set procWindows to windows of proc
            if (count of procWindows) > 0 then
              set win to item 1 of procWindows
              set winPos to position of win
              set winSize to size of win
              set winBounds to ((item 1 of winPos) as string) & "," & ((item 2 of winPos) as string) & "," & ((item 1 of winSize) as string) & "," & ((item 2 of winSize) as string)
            end if
          end try
          set appList to appList & procName & "|" & procBundle & "|" & procPID & "|" & winBounds & "\\n"
        end repeat
        return appList
      end tell
    `;

    const result = await runAppleScript(script);
    const apps: AppInfo[] = [];

    for (const line of result.split('\n')) {
      if (!line.trim()) continue;
      const [name, bundleId, pidStr, boundsStr] = line.split('|');
      if (!name || !bundleId) continue;

      const [x, y, width, height] = (boundsStr || '0,0,0,0').split(',').map(Number);
      apps.push({
        name: name.trim(),
        bundleId: bundleId.trim(),
        pid: parseInt(pidStr) || 0,
        bounds: { x: x || 0, y: y || 0, width: width || 0, height: height || 0 }
      });
    }

    return apps;
  }

  async focusApplication(identifier: string): Promise<boolean> {
    const script = `
      tell application "System Events"
        set targetProc to null
        repeat with proc in (every application process)
          if bundle identifier of proc is "${identifier}" or name of proc is "${identifier}" then
            set targetProc to proc
            exit repeat
          end if
        end repeat
        if targetProc is not null then
          set frontmost of targetProc to true
          return "success"
        else
          return "not found"
        end if
      end tell
    `;

    const result = await runAppleScript(script);
    if (result === 'success') {
      // Update currentApp
      const apps = await this.listApplications();
      this.currentApp = apps.find(a => a.bundleId === identifier || a.name === identifier) || null;
      return true;
    }
    return false;
  }

  async screenshot(options: { fullPage?: boolean; padding?: number } = {}): Promise<string> {
    const padding = options.padding || 10;

    // Take screenshot
    const screenshotBuffer = await screenshotDesktop({ format: 'png' });

    // If we have a focused app, crop to its window
    if (this.currentApp && this.currentApp.bounds.width > 0) {
      const { x, y, width, height } = this.currentApp.bounds;
      const cropped = await sharp(screenshotBuffer)
        .extract({
          left: Math.max(0, x - padding),
          top: Math.max(0, y - padding),
          width: width + (padding * 2),
          height: height + (padding * 2)
        })
        .png()
        .toBuffer();
      return cropped.toString('base64');
    }

    return screenshotBuffer.toString('base64');
  }

  async click(x: number, y: number, button: 'left' | 'right' = 'left'): Promise<boolean> {
    if (!this.currentApp) {
      throw new Error('No application focused. Call focusApplication first.');
    }

    const { bounds } = this.currentApp;
    const absX = bounds.x + (x * bounds.width);
    const absY = bounds.y + (y * bounds.height);

    await mouse.setPosition({ x: Math.round(absX), y: Math.round(absY) });
    await mouse.click(button === 'right' ? Button.RIGHT : Button.LEFT);
    return true;
  }

  async getClickableElements(): Promise<any[]> {
    if (!this.currentApp) {
      return [];
    }
    return await this.appleWindowManager.getClickableElements(this.currentApp.name);
  }

  async typeText(text: string): Promise<boolean> {
    await keyboard.type(text);
    return true;
  }

  async pressKey(key: string): Promise<boolean> {
    const keyMap: Record<string, Key> = {
      'Enter': Key.Enter,
      'Tab': Key.Tab,
      'Escape': Key.Escape,
      'Backspace': Key.Backspace,
      'Delete': Key.Delete,
      'ArrowUp': Key.Up,
      'ArrowDown': Key.Down,
      'ArrowLeft': Key.Left,
      'ArrowRight': Key.Right,
      'Space': Key.Space,
    };

    const nutKey = keyMap[key];
    if (nutKey) {
      await keyboard.pressKey(nutKey);
      await keyboard.releaseKey(nutKey);
    }
    return true;
  }

  async analyzeWithOCR(): Promise<any> {
    const screenshotBuffer = await screenshotDesktop({ format: 'png' });
    return await this.ocrAnalyzer.analyzeImage(screenshotBuffer);
  }

  async checkPermissions(): Promise<{ hasPermission: boolean; error?: string }> {
    return await checkAccessibilityPermissions();
  }

  createServer(): http.Server {
    return http.createServer(async (req, res) => {
      const url = new URL(req.url || '/', `http://localhost:${PORT}`);
      const reqPath = url.pathname;

      // CORS headers
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
      res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
      res.setHeader('Content-Type', 'application/json');

      // Handle preflight
      if (req.method === 'OPTIONS') {
        res.writeHead(200);
        res.end();
        return;
      }

      // API key authentication (except for health check)
      if (reqPath !== '/health') {
        const authHeader = req.headers['authorization'];
        const providedKey = authHeader?.replace('Bearer ', '');

        if (!providedKey || providedKey !== this.apiKey) {
          res.writeHead(401);
          res.end(JSON.stringify({ error: 'Unauthorized. Include Authorization: Bearer <api-key> header.' }));
          return;
        }
      }

      try {
        let body = '';
        if (req.method === 'POST') {
          for await (const chunk of req) {
            body += chunk;
          }
        }
        const params = body ? JSON.parse(body) : {};

        let result: any;

        switch (reqPath) {
          case '/health':
            result = { status: 'ok', version: '1.1.15' };
            break;

          case '/permissions':
            result = await this.checkPermissions();
            break;

          case '/listApplications':
            result = await this.listApplications();
            break;

          case '/focusApplication':
            if (!params.identifier) {
              throw new Error('identifier is required');
            }
            result = await this.focusApplication(params.identifier);
            break;

          case '/screenshot':
            const base64 = await this.screenshot(params);
            result = { image: base64, format: 'png' };
            break;

          case '/click':
            if (params.x === undefined || params.y === undefined) {
              throw new Error('x and y coordinates are required');
            }
            result = await this.click(params.x, params.y, params.button);
            break;

          case '/getClickableElements':
            result = await this.getClickableElements();
            break;

          case '/typeText':
            if (!params.text) {
              throw new Error('text is required');
            }
            result = await this.typeText(params.text);
            break;

          case '/pressKey':
            if (!params.key) {
              throw new Error('key is required');
            }
            result = await this.pressKey(params.key);
            break;

          case '/analyzeWithOCR':
            result = await this.analyzeWithOCR();
            break;

          case '/currentApp':
            result = this.currentApp;
            break;

          default:
            res.writeHead(404);
            res.end(JSON.stringify({ error: 'Not found', path: reqPath }));
            return;
        }

        res.writeHead(200);
        res.end(JSON.stringify(result));
      } catch (error: any) {
        console.error(`Error handling ${reqPath}:`, error);
        res.writeHead(500);
        res.end(JSON.stringify({ error: error.message }));
      }
    });
  }

  async start(): Promise<void> {
    // Check permissions first
    console.log('🎯 MCP-Eyes HTTP Server');
    console.log('=======================');
    console.log('');

    const permissions = await this.checkPermissions();
    if (permissions.hasPermission) {
      console.log('✅ Accessibility permissions: OK');
    } else {
      console.log('⚠️  Accessibility permissions: MISSING');
      console.log('   Grant Accessibility permission to iTerm/Terminal in System Settings');
    }

    console.log('');
    console.log('🔐 Security:');
    console.log(`   API Key: ${this.apiKey.substring(0, 8)}...${this.apiKey.substring(this.apiKey.length - 8)}`);
    console.log(`   Token file: ${TOKEN_FILE}`);
    console.log('   All requests require: Authorization: Bearer <api-key>');

    const server = this.createServer();
    server.listen(PORT, HOST, () => {
      console.log(`\n🚀 Server running at http://${HOST}:${PORT}`);
      console.log('\nEndpoints:');
      console.log('  GET  /health              - Health check (no auth required)');
      console.log('  GET  /permissions         - Check accessibility permissions');
      console.log('  GET  /listApplications    - List running apps with window bounds');
      console.log('  POST /focusApplication    - Focus an app (body: {identifier})');
      console.log('  POST /screenshot          - Take screenshot (body: {padding?})');
      console.log('  POST /click               - Click at position (body: {x, y, button?})');
      console.log('  GET  /getClickableElements - Get clickable UI elements');
      console.log('  POST /typeText            - Type text (body: {text})');
      console.log('  POST /pressKey            - Press key (body: {key})');
      console.log('  GET  /analyzeWithOCR      - Analyze screen with OCR');
      console.log('  GET  /currentApp          - Get currently focused app');
      console.log('\nPress Ctrl+C to stop');
    });
  }
}

// Start server
const server = new MCPEyesHTTPServer();
server.start().catch(console.error);

File: native/src/http_server.cpp
#include "mcp_eyes.h"
#include "platform.h"
#include <httplib.h>
#include <nlohmann/json.hpp>
#include <thread>
#include <atomic>
#include <mutex>

namespace mcp_eyes {

using json = nlohmann::json;

// External: web UI HTML (embedded or loaded from file)
extern const char* get_web_ui_html();

class HttpServer {
public:
    HttpServer(Agent* agent, Platform* platform)
        : agent_(agent), platform_(platform), running_(false) {}

    bool start(const AgentConfig& config) {
        if (running_) return false;

        config_ = config;

        // Determine bind address
        std::string bind_addr = "127.0.0.1";  // localhost default
        if (config_.network_mode == "lan" || config_.network_mode == "wan") {
            bind_addr = "0.0.0.0";
        }

        server_ = std::make_unique<httplib::Server>();
        setup_routes();

        running_ = true;
        server_thread_ = std::thread([this, bind_addr]() {
            server_->listen(bind_addr, config_.port);
        });

        return true;
    }

    void stop() {
        if (!running_) return;
        running_ = false;
        server_->stop();
        if (server_thread_.joinable()) {
            server_thread_.join();
        }
    }

    bool is_running() const { return running_; }

private:
    void setup_routes() {
        // CORS middleware
        server_->set_pre_routing_handler([](const httplib::Request& req, httplib::Response& res) {
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
            res.set_header("Access-Control-Allow-Headers", "Content-Type, Authorization");
            if (req.method == "OPTIONS") {
                res.status = 200;
                return httplib::Server::HandlerResponse::Handled;
            }
            return httplib::Server::HandlerResponse::Unhandled;
        });

        // Auth middleware (skip for health and web UI)
        auto auth_check = [this](const httplib::Request& req) -> bool {
            if (config_.network_mode == "localhost") return true;

            auto auth = req.get_header_value("Authorization");
            if (auth.empty()) return false;

            // Bearer token
            if (auth.substr(0, 7) == "Bearer ") {
                return auth.substr(7) == config_.api_key;
            }
            return false;
        };

        // ═══════════════════════════════════════════════════════════════
        // Web UI
        // ═══════════════════════════════════════════════════════════════

        server_->Get("/", [](const httplib::Request&, httplib::Response& res) {
            res.set_content(get_web_ui_html(), "text/html");
        });

        // ═══════════════════════════════════════════════════════════════
        // Health & Status (no auth required)
        // ═══════════════════════════════════════════════════════════════

        server_->Get("/health", [this](const httplib::Request&, httplib::Response& res) {
            json j = {{"status", "ok"}, {"version", VERSION}};
            res.set_content(j.dump(), "application/json");
        });

        server_->Get("/api/status", [this](const httplib::Request&, httplib::Response& res) {
            auto status = agent_->status();
            json j = {
                {"name", status.name},
                {"version", status.version},
                {"os", status.os},
                {"os_version", status.os_version},
                {"arch", status.arch},
                {"hostname", status.hostname},
                {"port", status.port},
                {"tls", status.tls},
                {"uptime_seconds", status.uptime_seconds},
                {"permissions", {
                    {"accessibility", status.permissions.accessibility},
                    {"screen_recording", status.permissions.screen_recording},
                    {"automation", status.permissions.automation}
                }}
            };
            res.set_content(j.dump(), "application/json");
        });

        // ═══════════════════════════════════════════════════════════════
        // Settings API
        // ═══════════════════════════════════════════════════════════════

        server_->Get("/api/settings", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            auto& config = agent_->config();
            json j = {
                {"name", config.name},
                {"network_mode", config.network_mode},
                {"port", config.port},
                {"tls_enabled", config.tls_enabled},
                {"api_key", config.api_key},
                {"allowed_ips", config.allowed_ips}
            };
            res.set_content(j.dump(), "application/json");
        });

        server_->Post("/api/settings", [this](const httplib::Request& req, httplib::Response& res) {
            try {
                auto j = json::parse(req.body);
                auto& config = agent_->config();

                if (j.contains("name")) config.name = j["name"];
                if (j.contains("network_mode")) config.network_mode = j["network_mode"];
                if (j.contains("tls_enabled")) config.tls_enabled = j["tls_enabled"];
                if (j.contains("allowed_ips")) config.allowed_ips = j["allowed_ips"].get<std::vector<std::string>>();

                agent_->save_config();
                res.set_content(R"({"success": true})", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/api/regenerate-key", [this](const httplib::Request&, httplib::Response& res) {
            // Generate new API key
            // ... implementation
            res.set_content(R"({"success": true})", "application/json");
        });

        // ═══════════════════════════════════════════════════════════════
        // MCP Operations (require auth in network mode)
        // ═══════════════════════════════════════════════════════════════

        server_->Get("/permissions", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            auto perms = platform_->check_permissions();
            json j = {
                {"accessibility", perms.accessibility},
                {"screen_recording", perms.screen_recording},
                {"automation", perms.automation}
            };
            res.set_content(j.dump(), "application/json");
        });

        server_->Get("/listApplications", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            auto apps = platform_->list_applications();
            json j = json::array();
            for (const auto& app : apps) {
                json windows = json::array();
                for (const auto& window : app.windows) {
                    windows.push_back({
                        {"title", window.title},
                        {"bounds", {
                            {"x", window.bounds.x},
                            {"y", window.bounds.y},
                            {"width", window.bounds.width},
                            {"height", window.bounds.height}
                        }},
                        {"is_minimized", window.is_minimized},
                        {"is_main", window.is_main}
                    });
                }

                j.push_back({
                    {"name", app.name},
                    {"bundle_id", app.bundle_id},
                    {"pid", app.pid},
                    {"bounds", {
                        {"x", app.bounds.x},
                        {"y", app.bounds.y},
                        {"width", app.bounds.width},
                        {"height", app.bounds.height}
                    }},
                    {"windows", windows}
                });
            }
            res.set_content(j.dump(), "application/json");
        });

        server_->Post("/focusApplication", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string identifier = j["identifier"];
                bool success = platform_->focus_application(identifier);
                res.set_content(success ? "true" : "false", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/screenshot", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                int padding = 0;
                if (!req.body.empty()) {
                    auto j = json::parse(req.body);
                    if (j.contains("padding")) padding = j["padding"];
                }

                auto* focused = platform_->get_focused_app();
                auto screenshot = platform_->take_screenshot(focused, padding);

                // Base64 encode
                static const char* b64_table = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
                std::string base64;
                // ... base64 encoding implementation

                json j = {{"image", base64}, {"format", "png"}};
                res.set_content(j.dump(), "application/json");
            } catch (const std::exception& e) {
                res.status = 500;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/click", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                float x = j["x"];
                float y = j["y"];
                bool right = j.value("button", "left") == "right";

                auto* focused = platform_->get_focused_app();
                if (!focused) {
                    res.status = 400;
                    res.set_content(R"({"error": "No app focused"})", "application/json");
                    return;
                }

                // Convert normalized to absolute
                int abs_x = focused->bounds.x + static_cast<int>(x * focused->bounds.width);
                int abs_y = focused->bounds.y + static_cast<int>(y * focused->bounds.height);

                bool success = platform_->click(abs_x, abs_y, right);
                res.set_content(success ? "true" : "false", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/click_absolute", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                int x = j["x"];
                int y = j["y"];
                bool right = j.value("button", "left") == "right";

                bool success = platform_->click(x, y, right);
                res.set_content(success ? "true" : "false", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/typeText", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string text = j["text"];
                bool success = platform_->type_text(text);
                res.set_content(success ? "true" : "false", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/pressKey", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string key = j["key"];
                bool success = platform_->press_key(key);
                res.set_content(success ? "true" : "false", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Get("/getClickableElements", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            auto* focused = platform_->get_focused_app();
            if (!focused) {
                res.set_content("[]", "application/json");
                return;
            }

            auto elements = platform_->get_clickable_elements(focused->name, true);
            json j = json::array();
            for (const auto& el : elements) {
                j.push_back({
                    {"type", el.type},
                    {"text", el.text},
                    {"role", el.role},
                    {"bounds", {
                        {"x", el.bounds.x},
                        {"y", el.bounds.y},
                        {"width", el.bounds.width},
                        {"height", el.bounds.height}
                    }},
                    {"normalized_position", {
                        {"x", el.normalized_position.x},
                        {"y", el.normalized_position.y}
                    }},
                    {"is_clickable", el.is_clickable},
                    {"is_enabled", el.is_enabled}
                });
            }
            res.set_content(j.dump(), "application/json");
        });

        server_->Get("/getUIElements", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            auto* focused = platform_->get_focused_app();
            if (!focused) {
                res.set_content("[]", "application/json");
                return;
            }

            auto elements = platform_->get_clickable_elements(focused->name, false);
            json j = json::array();
            for (const auto& el : elements) {
                j.push_back({
                    {"type", el.type},
                    {"text", el.text},
                    {"role", el.role},
                    {"bounds", {
                        {"x", el.bounds.x},
                        {"y", el.bounds.y},
                        {"width", el.bounds.width},
                        {"height", el.bounds.height}
                    }},
                    {"normalized_position", {
                        {"x", el.normalized_position.x},
                        {"y", el.normalized_position.y}
                    }},
                    {"is_clickable", el.is_clickable},
                    {"is_enabled", el.is_enabled}
                });
            }
            res.set_content(j.dump(), "application/json");
        });

        server_->Get("/analyzeWithOCR", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto screenshot = platform_->take_screenshot();
                auto results = platform_->perform_ocr(screenshot);

                json list = json::array();
                for (const auto& r : results) {
                    list.push_back({
                        {"text", r.text},
                        {"confidence", r.confidence},
                        {"bounds", {
                            {"x", r.bounds.x},
                            {"y", r.bounds.y},
                            {"width", r.bounds.width},
                            {"height", r.bounds.height}
                        }}
                    });
                }
                json payload = {
                    {"width", screenshot.width},
                    {"height", screenshot.height},
                    {"results", list}
                };
                res.set_content(payload.dump(), "application/json");
            } catch (const std::exception& e) {
                res.status = 500;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        // ═══════════════════════════════════════════════════════════════
        // High-level helper endpoints
        // ═══════════════════════════════════════════════════════════════

        server_->Post("/openUrl", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string url = j["url"];

                // Platform-specific URL open
                #ifdef __APPLE__
                    system(("open \"" + url + "\"").c_str());
                #elif _WIN32
                    system(("start \"\" \"" + url + "\"").c_str());
                #else
                    system(("xdg-open \"" + url + "\"").c_str());
                #endif

                res.set_content(R"({"success": true})", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/openApp", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string app_name = j["app"];

                // Platform-specific app launch
                #ifdef __APPLE__
                    system(("open -a \"" + app_name + "\"").c_str());
                #elif _WIN32
                    system(("start \"\" \"" + app_name + "\"").c_str());
                #else
                    system((app_name + " &").c_str());
                #endif

                res.set_content(R"({"success": true})", "application/json");
            } catch (const std::exception& e) {
                res.status = 400;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Post("/findAndClick", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto j = json::parse(req.body);
                std::string text = j["text"];

                // Take screenshot, OCR, find text, click
                auto screenshot = platform_->take_screenshot();
                auto results = platform_->perform_ocr(screenshot);

                for (const auto& r : results) {
                    if (r.text.find(text) != std::string::npos) {
                        // Click center of found text
                        int click_x = r.bounds.x + r.bounds.width / 2;
                        int click_y = r.bounds.y + r.bounds.height / 2;
                        platform_->click(click_x, click_y, false);

                        json result = {{"success", true}, {"clicked_text", r.text}};
                        res.set_content(result.dump(), "application/json");
                        return;
                    }
                }

                json result = {{"success", false}, {"error", "Text not found"}};
                res.set_content(result.dump(), "application/json");
            } catch (const std::exception& e) {
                res.status = 500;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });

        server_->Get("/readScreen", [this, auth_check](const httplib::Request& req, httplib::Response& res) {
            if (!auth_check(req)) { res.status = 401; return; }

            try {
                auto screenshot = platform_->take_screenshot();
                auto results = platform_->perform_ocr(screenshot);

                std::string all_text;
                for (const auto& r : results) {
                    all_text += r.text + "\n";
                }

                json j = {{"text", all_text}};
                res.set_content(j.dump(), "application/json");
            } catch (const std::exception& e) {
                res.status = 500;
                json err = {{"error", e.what()}};
                res.set_content(err.dump(), "application/json");
            }
        });
    }

    Agent* agent_;
    Platform* platform_;
    AgentConfig config_;
    std::unique_ptr<httplib::Server> server_;
    std::thread server_thread_;
    std::atomic<bool> running_;
};

} // namespace mcp_eyes
